Added run_id to run_context implementation.

joakim-hove opened the pull request:
**Task**
Added a run_id to the `ert_run_context`

**Approach**
Use time/uid/pid and some random numbers to create a random ID.

**Pre un-WIP checklist**
- [x] Statoil tests pass locally
- [x] Have completed graphical integration test steps

**Depends on**
* TNO/Seba#109
* Statoil/ert#44
* https://ci.tno.nl/gitlab/statoil/Seba/merge_requests/38

Observe that this is *tightly bound* to https://github.com/TNO/Seba/pull/109 - the commits marked with: `{libres:xxx}` should have a matching commit in the SEBA PR.

**Extended description**

When starting a bunch of simulations there is a set of metadata which describes this batch. The metadata include:

1. Cases:
   - Which case should be used for initialization.
   - Which case should be used to store the simulation results.
   - Which case should be used for updated parameters.
2. Which iteration number is this.
3. Which realisations are active.
4. Where should the realisation folders be located.
5. How should we initialize the realisations.
6. ....

All of this information is kept in an instance of `ert_run_context`, and the new `run_id` was added to this structure. The `ert_run_context` instance should have lifetime which can be illustrated with the following hypothetical `with`statement:

```python
with ErtRunContext(  ) as run_context:
    init_realisations( run_context )
    create_folders( run_context )
    submit:_simulations( run_context )
    update_realisations( run_context )
``` 

However it turned out the actual use of the `ert_run_context` type was not exactly like this; on the contrary `ert_run_context` instances were recreated from relevant underlying arguments several times one logical initialize-simulate-store-update cycle, with the consequence that the `run_id` became completely bogus. 

So the bulk of this is many changes to ascertain that the `ert_run_context` instance has the correct lifetime. Unfortunately the most fragile (i.e. not-really-tested) changes are in the gui code in Statoil/ert#44 and also the changes in Seba are 100% untested.

joakim-hove committed:
Added run_id to run_context implementation.

joakim-hove committed:
Added run_id argument to run_arg.

joakim-hove committed:
'simulation_id' -> 'run_id' in FORWARD_MODEL.

joakim-hove committed:
Pass run_id to the forward model output.

epa095 commented on libenkf/src/ert_run_context.c:
```diff
@@ -86,7 +90,15 @@ stringlist_type * ert_run_context_alloc_runpath_list(const bool_vector_type * ia
+  time_t now = time( NULL );
+  unsigned int random = util_dev_urandom_seed( );
+  util_set_datetime_values_utc( now , &sec, &min, &hour, &day, &month, &year);
+  return util_alloc_sprintf("%d:%d:%4d-%0d-%02d-%02d-%02d-%02d:%ud" , getpid() , getuid(), year , month , day , hour , min , sec, random);
```
I must admit that I get a bad feeling about embedding this information this clearly/nicely in the ID. Not that I have a problem with them being the _source_ of the ID, but when they are so nicely formatted it is easy for someone (e.g. me in the future) to think "I would like to have the uid, ohh it is right there in the ID, let me just extract it" and suddenly the ID is no longer just a id, it is actually a structure containing more information, and we are locked to keep it like that forever, or figure out who (a)buses the field and change them. I would feel better if we hashed or scrambled the data a bit first. 

If we are interested in the pid/uid/local-timestamp we should send them as explicit fields. I have no problems with that, there is no reason to be stingy with the meta-information we send in the jobs-document, as long as it is explicit.

joakim-hove replied:
> [....] I would like to have the uid, ohh it is right there in the ID, let me just extract it

I thought of that too - and I agree with you. Will change it.

joakim-hove replied:
I have not changed this yet; all in all the implementation of the `run_id` became a completely insignificant part of this PR. I still agree with @epa095 and will change the format of the `run_id`- but would like to merge this first.

markusdregi replied:
I've made a Jira task regarding the format of the `run_id` in an attempt to prevent that this discussion goes mute when this PR is merged..

joakim-hove committed:
Removed ert_init_context - use ert_run_context

joakim-hove committed:
Removed target case arg {libres:rm_target_case}

joakim-hove committed:
Added target_case arg {libres:add_target_case}

joakim-hove committed:
Using ert_run_context in simulation_context.

joakim-hove committed:
Changing to use ert_run_context for init.

joakim-hove committed:
Create runpath can initialize case.

joakim-hove committed:
Changed numerous call sites to use run_context.

joakim-hove committed:
Removed write count from EnkfFS __repr__

markusdregi commented on libenkf/src/enkf_main.c:
```diff
@@ -1806,13 +1792,14 @@ void enkf_main_run_iterated_ES(enkf_main_type * enkf_main,
 
-ert_init_context_type * enkf_main_alloc_ert_init_context(const enkf_main_type * enkf_main , enkf_fs_type * fs, const bool_vector_type * iactive , init_mode_type init_mode , int iter) {
+
+/*ert_init_context_type * enkf_main_alloc_ert_init_context(const enkf_main_type * enkf_main , enkf_fs_type * fs, const bool_vector_type * iactive , init_mode_type init_mode , int iter) {
```
Please remove this completely.

markusdregi commented on libenkf/src/ert_run_context.c:
```diff
@@ -146,12 +191,39 @@ ert_run_context_type * ert_run_context_alloc_SMOOTHER_RUN(enkf_fs_type * simulat
+                                             path_fmt_type * runpath_fmt ,
+                                             subst_list_type * subst_list ,
+                                             int iter) {
+  if (run_mode == SMOOTHER_UPDATE)
```
Please consider using a switch for handling various enum values. The compiler will then help us keeping track of when a specific value is not handled.

markusdregi commented on python/python/res/enkf/ert_run_context.py:
```diff
@@ -69,3 +115,27 @@ def createRunpathList(cls, mask, runpath_fmt, subst_list, iter=0):
         return cls._alloc_runpath(iens, runpath_fmt, subst_list, iter)
+
+    
+    def get_id(self):
```
These all seems like good candidates for being properties..?

markusdregi commented on python/python/res/server/ertrpcclient.py:
```diff
@@ -13,6 +13,7 @@
 
 
 def convertFault(fault):
+    print fault
```
Remove?

markusdregi commented on python/python/res/server/simulation_context.py:
```diff
@@ -1,50 +1,53 @@
     def isRunning(self):
         return self._queue_manager.isRunning()
-
+    
```
Trailing whitespace, please remove

markusdregi commented on python/python/res/server/simulation_context.py:
```diff
@@ -93,3 +96,10 @@ def __repr__(self):
         fmt =  fmt % (running, numRunn, numSucc, numFail, numWait)
         return 'SimulationContext(%s)' % fmt
+
+    def get_result_fs(self):
```
To me it seems like `result_fs`and `run_context` are good candidates for being properties..

markusdregi commented on python/tests/res/enkf/test_runpath_list.py:
```diff
@@ -127,36 +128,44 @@ def test_assert_export(self):
+#            runpath_list = ert.getRunpathList( )
+#
+#            ens_size = ert.getEnsembleSize( )
+#            runner = ert.getEnkfSimulationRunner( )
```
Please uncomment test

markusdregi commented on python/tests/res/server/test_rpc_service.py:
```diff
@@ -40,7 +41,7 @@ def test_server_creation(self):
             server.stop()
 
-    def test_client_interaction(self):
+    def xtest_client_interaction(self):
```
Doesn't the naming `xtest..` result in this particular test never being ran?

joakim-hove replied:
Ehhh - yes it does; hopefully changing it back does not give a surprise ...

markusdregi commented on python/tests/res/server/test_rpc_service.py:
```diff
@@ -170,15 +174,19 @@ def clientInteraction():
             for thread in threads:
                 thread.start()
-
+            
```
Trailing whitespace

markusdregi commented on python/tests/res/server/test_rpc_storage.py:
```diff
@@ -74,7 +94,7 @@ def test_rpc_storage(self):
                 self.assertEqual(data["Test:GradientDirection"][sim_id], gradient_directions[sim_id])
 
-
+                
```
...

markusdregi commented on python/tests/res/server/test_rpc_storage.py:
```diff
@@ -84,25 +104,25 @@ def test_rpc_storage_with_simulation(self):
 
         for iens in range(realization_count):
-            client.addSimulation(batch_name, geo_id=0, pert_id=0, sim_id=iens, keywords=kw)
+            client.addSimulation(0,0,iens,  kw)
```
Please fix the white space between the arguments

markusdregi reviewed (approved):
This is excellent and highly appreciated work! It ended up having much bigger ramifications then first anticipated, but these are truly important.

Good job! Ship it when the comments are addressed.

joakim-hove committed:
Using switch for run_context alloc.

joakim-hove committed:
Uncommented test.

joakim-hove committed:
enkf_main_get_analysis_config: NOT const return.

joakim-hove committed:
ErtTestContext: will raise IOError.

The ErtTestContext class will raise IOError if you try to load a
non-existing job.

joakim-hove committed:
Changed import path to res.enkf in plugins.

joakim-hove committed:
Removed 'stale' test.

The test_client_interaction( ) test was extensively using the ability to
pass individual cases to the RPC addSimulation( ) method. That ability
no longer exists - a common case must be used for the whole simulation
batch, this test was therefor deemed to be beyond salvation.

pgdr commented on python/python/res/server/ertrpcserver.py:
```diff
@@ -137,23 +137,25 @@ def isInitializationCaseAvailable(self):
                 self._session.batch_number += 1
-                self._session.simulation_context = SimulationContext(self.ert, simulation_count, verbose=self._verbose_queue)
-
+                
```
Please don't introduce new trailing white spaces.

joakim-hove committed:
White space clean up
